accepted/cranelift-isel-pre-rfc.md

```diff
@@ -0,0 +1,798 @@
+
+## Requirement 2: Transition from Hand-Written Code
+
+A key fact to consider is that *our machine backends already exist*,
+and they must continue working as we implement and make use of
+whatever DSL we define. Furthermore, we should do as much as we can to
+avoid large atomic switchover PRs. It is much easier to (i) make
+progress and (ii) verify correctness as we go if we can migrate small
+bits of backend logic at a time.
+
+There are two ways in which DSL-written and hand-written code can
+interact:
+
+- "Horizontally", between different instructions. For example, we
+  might start to use a DSL to define arithmetic/SIMD instructions, and
+  optimize SIMD lowerings, while retaining the existing (working,
+  verified, hand-optimized) lowering for all other CLIF
+  instructions. Or we might want to prove out the DSL by fully
+  implementing all of the existing pattern-matching cases for just one
+  instruction. It should be possible to do this while deferring to
+  existing handwritten code for all not-handled cases.
+
+  This is the simplest form of co-existence and would allow a
+  migration strategy by which we implement the entire lowering logic
+  for one or more Cranelift opcodes at a time, eventually covering the
+  whole CLIF opcode space.
+
+- "Vertically", within the logic to lower a single CLIF
+  instruction. For example, we might use the DSL to define lowering
+  patterns for arithmetic instructions and combinations of arithmetic
+  instructions, but defer to existing handwritten code that looks for
+  loads from memory that can be merged into the arithmetic
+  instructions and computes the best addressing mode with more complex
+  logic. Or, going the other way, we might try to systematize the
+  addressing-mode lowering into a set of rules, and use it from
+  existing handwritten code that lowers loads/stores.
+  
+  This implies a sort of "FFI" by which the DSL code can call out to
+  Rust, and perhaps the Rust code can call back into (sub)rules in the
+  DSL.
+  
+  This is a more challenging form of interoperation of old and new
+  (though it may fall more naturally out of some approaches than
+  others). If we can achieve it, it would allow us to (i) continue to
+  make use of shared code, such as code that computes addressing
+  modes, various forms of immediates, etc., when writing new lowerings
+  in the DSL, and (ii) have more flexibility with certain complex
+  lowering logic. The main advantage can thus be described as
+  "flexibility".
+  
+  The main downside of "vertical" interop is that, by providing more
+  flexibility, it can make analyzability more difficult. For example,
+  if we aim to verify the lowerings, we need to annotate what these
+  calls to outside code are doing for the sake of the verification
+  tool.
+  
+  This is reminiscent of "unsafe" code in Rust: it allows one to build
+  axiomatic building blocks with flexibility, but it requires one to
+  carefully define *what* the building blocks do.
```

Comment:
I think that even after the migration is complete it should be possible to write arbitrary code for the output generation, but not necessarily the input pattern matching. This would allow for complex output code to handle for example turning fixed divisions into multiplications or shifts while still preserving analyzability and optimizability of the patterns if desired.

Reply:
Right, there are a number of cases where "little building block of arbitrary logic" is useful. Creation of immediate operands is another good example: aarch64 has a very interesting logical-immediate format that can support only some values, with a [complex algorithm](https://github.com/bytecodealliance/wasmtime/blob/bb85366a3bfab793de960b5be2b069760ab8a6df/cranelift/codegen/src/isa/aarch64/inst/imms.rs#L345-L535) to derive it. We'd want to make that a "primitive" in some sense by calling out to the existing implementation.